src/merkle_tree.rs

```diff
@@ -0,0 +1,1383 @@
+                    (node_hash, NodeType::Mid)
+                }
+            } else {
+                (child_hash, child_type)
+            }
+        }
+    } else if depth == 255 {
+        // this is an edge case where we make it all the way down to the
+        // bottom of the tree, and split the last pair. This has the same
+        // effect as the else-block, but since we use u8 for depth, it would
+        // overflow
+        debug_assert!(range.len() > 1);
+        debug_assert!(left < range.len() as i32);
+
+        merkle_tree.nodes_vec.push(ArrayTypes::Leaf {
+            data: merkle_tree.leaf_vec.len(),
+        });
+        merkle_tree.leaf_vec.push(range[0]);
+        merkle_tree.hash_cache.push(hash_leaf(range[0]));
+
+        merkle_tree.nodes_vec.push(ArrayTypes::Leaf {
+            data: merkle_tree.leaf_vec.len(),
+        });
+        merkle_tree.leaf_vec.push(range[left as usize]);
+        merkle_tree.hash_cache.push(hash_leaf(range[left as usize]));
+
+        let nodes_len = merkle_tree.nodes_vec.len();
+        merkle_tree.nodes_vec.push(ArrayTypes::Middle {
+            children: (nodes_len - 2, nodes_len - 1),
+        });
+        let node_hash = hash(
+            NodeType::Term,
+            NodeType::Term,
+            &range[0],
+            &range[left as usize],
+        );
+        merkle_tree.hash_cache.push(node_hash);
+        (node_hash, NodeType::MidDbl)
+    } else {
+        // we are a middle node
+        // recursively sort and hash our left and right children and return the resultant hash upwards
+        let (left_hash, left_type) =
+            generate_merkle_tree_recurse(&mut range[..left as usize], depth + 1, merkle_tree);
+        // make a note of where the left child node is
+        let left_child_index: usize = merkle_tree.nodes_vec.len() - 1;
+        let (right_hash, right_type) =
+            generate_merkle_tree_recurse(&mut range[left as usize..], depth + 1, merkle_tree);
+
+        let node_type: NodeType = if left_type == NodeType::Term && right_type == NodeType::Term {
+            // Prune the two empties beneath us and push ourselves as empty
+            // merkle_tree.nodes_vec.remove(usize::try_from(left_child_index).unwrap());
+            // merkle_tree.nodes_vec.remove(merkle_tree.nodes_vec.len() - 1);
+            // merkle_tree.nodes_vec.push(ArrayTypes::Empty);
+            merkle_tree.nodes_vec.push(ArrayTypes::Middle {
+                children: (left_child_index, merkle_tree.nodes_vec.len() - 1),
+            });
+            NodeType::MidDbl
+        } else {
+            merkle_tree.nodes_vec.push(ArrayTypes::Middle {
+                children: (left_child_index, merkle_tree.nodes_vec.len() - 1),
+            });
+            NodeType::Mid
+        };
+        let node_hash = hash(left_type, right_type, &left_hash, &right_hash);
+        merkle_tree.hash_cache.push(node_hash);
+        (node_hash, node_type)
+    }
+}
+
+
+#[cfg(test)]
```

Comment:
I typically put everything used in tests inside their own module, like this:
```rs
#[cfg(test)]
mod tests {
    use super::*;
    
    ...
}
```

That way you don't need to manually mark everything as `#[cfg(test)]` and you can import things naturally inside of this module.